### ISSUE TITLE
WebTransport/WebCodecs interaction (copies)

### ISSUE THREAD

**Issue by aboba** [w3c/webtransport#231]
There are a number of use cases where it is envisaged that WebTransport + WebCodecs will be used together: 

1. Video upload, where a captured MediaStreamTrack is converted to raw video frames, encoded by WebCodecs, and then sent to a server using WebTransport. 
2. Video streaming, where a video stream is sent to a WebTransport client, encodedVideoChunks are decoded using WebCodecs into videoFrames, converted to a MediaStreamTrack and rendered by a video tag. 

In use case #1, it is desirable for the encodedVideoChunk produced by WebCodecs to be passed to WebTransport and sent with as few copies as possible (e.g. just one copy for process separation). Is this possible? 

In use case #2, it is desirable for WebTransport datagram or reliable stream reader to transfer datagrams/segments of the encodedVideoChunks into a GPU buffer rather than main memory, so as to allow most efficient ingestion by WebCodecs and avoid a copy.  Is this possible?

**Comment by chcunningham** [comment]
I think you nailed the use cases, but the assumptions about GPU buffers aren't quite right (I probably didn't communicate this well in our meeting). 

Here's some background:
1. In Chrome, GPU memory would only ever be used to back a VideoFrame. This is generally done because VideoFrames ultimately get composited/rendered into GPU memory, so having them there to start with reduces CPU->GPU copies. 
2. We do use shared memory (but not GPU memory) when moving encoded chunks between processes, but this just an implementation detail.
3. For encoded chunks, we're generally not super concerned with copies because the data is relatively small. Chrome's implementation of existing <video> pipelines (e.g. MSE, WebRTC) makes a few copies encoded data, even when we use shared memory to do hardware accelerated decoding. 

In spite of point 3, I think we should still map out the flow from WebTransport to what copies of encoded data are forced by the current API shape.

**For use case #1 (sending encoded data -> web):**
- EncodedVideoChunk currently has a bug where it directly exposes its array buffer.  We should replace this with a copyInto() method (see https://github.com/w3c/webcodecs/issues/127 for justficiation). This means at least one copy from WC chunk going to WebTransport.
- I am not a streams expert, but I assume that we can write the entire ArrayBuffer to the WT stream in one pass, so no additional copies there. @aboba, sound right?

**For use case #2 (downloading encoded data from webtransport)
- I don't know offhand how much control WebTransport users have over the size of stream chunks (not to be confused with WC chunks) that they read from the ReadableStream. If control is limited, it may occur that the stream chunk is smaller than the full WC chunk such that additional buffering (copies) is required before being able to feed WC. 
- Once you have a full WC chunks worth of data, you can pass that as ArrayBuffer to create the WC chunk. Presently this copies the ArraysBuffer to prevent TOCTU issues. But we https://github.com/w3c/webcodecs/issues/104 proposes letting callers transfer this ArrayBuffer into the chunk.

Again, I don't think any of these copies is a big deal. Just trying to map it all out and look for optimizations.

**Comment by aboba** [comment]
@chcunningham There may be some differences depending which WebTransport mode is used.  WebTransport supports ordered/reliable transport (where data is sent over a single reliable stream), unordered/reliable transport (a distinct reliable stream for each message), or unreliable/unordered (HTTP/3 datagrams).  Reliable ordered transport on a single stream might be used with HLS or DASH, with WebCodecs substituted for MSE. For low-latency game streaming, you might want a different transport to minimize latency, such as unreliable/unordered (transporting packetized media over HTTP/3 datagrams, with the application handling robustness via its own RTX/FEC/RED), or partially reliable/unordered (e.g. using a distinct stream for each EncodedVideoChunk, with a retransmission time limit), or maybe some mixture (e.g. reliable stream for key frames, unordered/unreliable for P-frames).  

Let me try to walk through the use cases, covering reliable/ordered or unreliable/unordered transport.  I believe that the reliable/unordered case is similar to reliable/ordered. 

For use case #1, if I read Issue https://github.com/w3c/webcodecs/issues/155 correctly, if the goal is to upload containerized media, you could call writer.write(EncodedVideoChunk).  For datagrams you most likely would not be sending containerized media, so if the EncodedVideoChunk is provided in containerized form, it would be necessary to de-containerize it and then packetize. It would be nice to be able to select portions of the EncodedVideoChunk to be written into the datagram payloads without having to copy, so the de-containerization/packetization process can be most efficient.

For use case #2, if you are reading the EncodedVideoChunk from a reliable stream, and it is received in containerized form, you'd want the segments to be deposited into a contiguous ArrayBuffer which you can then feed to the WebCodecs decoder. For datagrams, if the media was not containerized for transport, but needs to be provided to WebCodecs in containerized form, you'd need to de-packetize and containerize the media before feeding it to the WebCodecs decoder.

**Comment by wilaw** [comment]
@aboba  - can you reference the discussions you spoke about in this mornings call? Concerning BYOB readers.

**Comment by yutakahirano** [comment]
Are both of datagrams and reliable streams target?

**Comment by aboba** [comment]
@yutakahirano Yes. Some use cases:
1. Video upload: WebCodecs encoding and sending via WebTransport reliable streams.  
2. Low-latency streaming:  WebTransport receiving unordered/unreliable datagrams, decoding via WebCodecs. 
3. Conventional streaming: WebTransport receiving reliable stream, decoding via WebCodecs.

**Comment by aboba** [comment]
@wilaw  Here are some references:
https://github.com/whatwg/streams/issues/495
https://github.com/w3c/webtransport/issues/131

**Comment by yutakahirano** [comment]
Here is my mental model for the reliable streams case, client => server.

1. A script provides the data to upload as a Uint8Array.
2. The browser copies the data into an IPC buffer __[A]__.
   - There may be one extra copy here, when the IPC buffer is full, but that can be avoidable with checking `writer.ready`.
3. In the network process, create QUIC frames for the data __[B]__.
4. Send the data to the network __[C]__.

__[A]__, __[B]__, __[C]__ are potential copies.

__[A]__ comes from the followings in the spec side:
 1. We have separate renderers and the network process. This is for security and (implicitly?) required by the spec.
 1. There is no way to write-and-transfer data at [WritableStreamDefaultWriter.write](https://streams.spec.whatwg.org/#default-writer-write).
 1. There is no way to allocate an ArrayBuffer to be accessible from multiple processes.

On the other hand, we may be able to optimize away the copy with (all of) the followings:
  1. We have a conversion function from a VideoFrame to bytes, and it's provided by the browser.
  2. The VideoFrame is not exposed to scripts (or, having a lock system to restrict the access to the video memory, similar to [ReadableStream.getReader](https://streams.spec.whatwg.org/#rs-get-reader) is good too).

In other words, if the contents of the video memory is not accessible to scripts then it is possible to provide the GPU memory handle directly to the network process and eliminate the copy at __[A]__.

Reg: __[B]__ I expect at least one copy here because we have a TLS encryption process here. I can think of a system where we allocate a buffer backed by the network hardware, and there is only one copy involved (the input is the IPC buffer, and the output is the buffer backed by the network hardware, and all the HTTP/3 and QUIC protocol process is done between them) for __[B]__ and __[C]__ combined. @DavidSchinazi knows much more than me here. 

@aboba, does this make sense / is this useful?

**Comment by yutakahirano** [comment]
One correction:

> In other words, if the contents of the video memory is not accessible to scripts then it is possible to provide the GPU memory handle directly to the network process and eliminate the copy at [A].

I came to think this is problematic in terms of security. Running the conversion logic should be done in the renderer, not in the network process. Still, we should be able to eliminate the copy and intermediary buffer allocation.

**Comment by jan-ivar** [comment]
Meeting:
- @aboba ship-blocking for webCodecs but maybe not us.

**Comment by aboba** [comment]
BYOB support has been added as of M108. Since encoded chunks are much smaller than VideoFrames, copies of encoded chunks between WebCodecs and WebTransport are not as big an issue as concurrency in frame/stream transport, which appears to be achievable by removing `await` before `writer.write()` and `writer.close()` (`await writer.ready` is fine).

**Comment by wilaw** [comment]
@aboba  - can we close this issue?
